### ISSUE TITLE
improve no_effect lint to lint partial expressions that don't have an effect

### ISSUE THREAD

**Issue by oli-obk** [Manishearth/rust-clippy#750]
currently we don't lint the following expressions:

```rust
    get_number();
    Tuple(get_number());
    Struct { field: get_number() };
    Struct { ..get_struct() };
    Enum::Tuple(get_number());
    Enum::Struct { field: get_number() };
    5 + get_number();
    *&get_number();
    &get_number();
    (5, 6, get_number());
    box get_number();
    get_number()..;
    ..get_number();
    5..get_number();
    [42, get_number()];
    [42, 55][get_number() as usize];
    (42, get_number()).1;
    [get_number(); 55];
    [42; 55][get_number() as usize];
```

An allow by default lint could lint all of those, since a function result remains unused and isn't explicitly ignored with `let _ = `.

More complex but imo warn by default could be a lint that suggests to simplify all of the above to just `get_number()`
